specs/_features/eip7594/das-core.md

```diff
@@ -0,0 +1,262 @@
+    return FlatExtendedMatrix(matrix)
+```
+
+#### `get_data_column_sidecars`
+
+```python
+def get_data_column_sidecars(signed_block: SignedBeaconBlock,
+                             blobs: Sequence[Blob]) -> Sequence[DataColumnSidecar]:
+    signed_block_header = compute_signed_block_header(signed_block)
+    block = signed_block.message
+    kzg_commitments_inclusion_proof = compute_merkle_proof(
+        block.body,
+        get_generalized_index(BeaconBlockBody, 'blob_kzg_commitments'),
+    )
+    cells_and_proofs = [compute_cells_and_proofs(blob) for blob in blobs]
+    blob_count = len(blobs)
+    cells = [cells_and_proofs[i][0] for i in range(blob_count)]
+    proofs = [cells_and_proofs[i][1] for i in range(blob_count)]
+    sidecars = []
+    for column_index in range(NUMBER_OF_COLUMNS):
+        column = DataColumn([cells[row_index][column_index]
+                             for row_index in range(blob_count)])
+        kzg_proof_of_column = [proofs[row_index][column_index]
+                               for row_index in range(blob_count)]
+        sidecars.append(DataColumnSidecar(
+            index=column_index,
+            column=column,
+            kzg_commitments=block.body.blob_kzg_commitments,
+            kzg_proofs=kzg_proof_of_column,
+            signed_block_header=signed_block_header,
+            kzg_commitments_inclusion_proof=kzg_commitments_inclusion_proof,
+        ))
+    return sidecars
+```
+
+## Custody
+
+### Custody requirement
+
+Each node downloads and custodies a minimum of `CUSTODY_REQUIREMENT` subnets per slot. The particular columns that the node is required to custody are selected pseudo-randomly (more on this below).
+
+A node *may* choose to custody and serve more than the minimum honesty requirement. Such a node explicitly advertises a number greater than `CUSTODY_REQUIREMENT` via the peer discovery mechanism -- for example, in their ENR (e.g. `custody_lines: 4` if the node custodies `4` subnets each slot) -- up to a `DATA_COLUMN_SIDECAR_SUBNET_COUNT` (i.e. a super-full node).
+
+A node stores the custodied columns for the duration of the pruning period and responds to peer requests for samples on those columns.
+
+### Public, deterministic selection
+
+The particular columns that a node custodies are selected pseudo-randomly as a function (`get_custody_columns`) of the node-id and custody size -- importantly this function can be run by any party as the inputs are all public.
+
+*Note*: increasing the `custody_size` parameter for a given `node_id` extends the returned list (rather than being an entirely new shuffle) such that if `custody_size` is unknown, the default `CUSTODY_REQUIREMENT` will be correct for a subset of the node's custody.
+
+## Peer discovery
+
+At each slot, a node needs to be able to readily sample from *any* set of columns. To this end, a node should find and maintain a set of diverse and reliable peers that can regularly satisfy their sampling demands.
```

Comment:
> At each slot, a node needs to be able to readily sample from *any* set of columns.

This includes the subnet that the node is assigned to custody right?

Reply:
Yes. You don't need to do anything when you already have a column you want to sample because you are in that subnet.

Reply:
Got it, thanks! So for nodes that subscribes to all column subnets, they won't be doing any sampling with its peers because they already have all the columns.

Reply:
For nodes custodying at least 50% of the columns, would they still need to sample peers?

Reply:
> For nodes custodying at least 50% of the columns, would they still need to sample peers?

I think they shouldn't because the data is now surely available.